pkg/controllers/streaming/processor_controller_test.go

```diff
@@ -17,44 +17,116 @@ limitations under the License.
+	"github.com/projectriff/system/pkg/tracker"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
 	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
-
-	"github.com/projectriff/system/pkg/apis"
-	"github.com/projectriff/system/pkg/apis/build/v1alpha1"
-	corev1alpha1 "github.com/projectriff/system/pkg/apis/core/v1alpha1"
-	streamingv1alpha1 "github.com/projectriff/system/pkg/apis/streaming/v1alpha1"
-	kedav1alpha1 "github.com/projectriff/system/pkg/apis/thirdparty/keda/v1alpha1"
-	rtesting "github.com/projectriff/system/pkg/controllers/testing"
-	"github.com/projectriff/system/pkg/refs"
-	"github.com/projectriff/system/pkg/tracker"
 )
 
-const (
-	testNamespace      = "test-namespace"
-	testName           = "test-processor"
-	testProcessorImage = "test-processor-image"
-	testDefaultImage   = "test-image-from-template"
-	testFunction       = "test-function"
-	testFunctionImage  = "test-function-image"
-	testContainer      = "test-container"
-	testContainerImage = "test-container-image"
-	testErrorMessage   = "test error"
-)
+func TestReconcile(t *testing.T) {
+	scheme := runtime.NewScheme()
+	_ = clientgoscheme.AddToScheme(scheme)
+	_ = corev1alpha1.AddToScheme(scheme)
+	_ = streamingv1alpha1.AddToScheme(scheme)
+	_ = kedav1alpha1.AddToScheme(scheme)
+	_ = v1alpha1.AddToScheme(scheme)
 
-var testError = errors.New(testErrorMessage)
+	const (
+		testNamespace      = "test-namespace"
+		testName           = "test-processor"
+		testProcessorImage = "test-processor-image"
+		testDefaultImage   = "test-image-from-template"
+		testFunction       = "test-function"
+		testFunctionImage  = "test-function-image"
+		testContainer      = "test-container"
+		testContainerImage = "test-container-image"
+	)
+
+	processorGiven := factories.Processor().
+		NamespaceName(testNamespace, testName).
+		PodTemplateSpec(func(spec factories.PodTemplateSpec) {
+			spec.ContainerNamed(testContainer, func(c *corev1.Container) {
+				c.Image = testDefaultImage
+			})
+		})
+	configMapGiven := factories.ConfigMap().
+		NamespaceName(testNamespace, processorImages).
+		AddData(processorImageKey, testProcessorImage)
+	containerGiven := factories.Container().
+		NamespaceName(testNamespace, testContainer).
+		StatusLatestImage(testContainerImage)
+	deploymentGiven := factories.Deployment().
+		ObjectMeta(func(om factories.ObjectMeta) {
+			om.Namespace(testNamespace)
+			om.GenerateName("%s-processor-", testName)
+			om.AddLabel("streaming.projectriff.io/processor", testName)
+			om.ControlledBy(processorGiven.Get(), scheme)
+		}).
+		Replicas(1).
+		AddSelectorLabel("streaming.projectriff.io/processor", testName).PodTemplateSpec(func(pts factories.PodTemplateSpec) {
+		pts.AddLabel("streaming.projectriff.io/processor", testName)
+	})
+	scaledObjectGiven := &kedav1alpha1.ScaledObject{
```

Comment:
@scothis The construction of this object is unlike most of the others. Do you think it would be worth introducing a factory (particular if scaled objects are going to crop up in others tests)?

Reply:
I created factories for kpack and knative types. Adding a factory for keda types makes sense. We can keep the factory simple and only expose the bits that vary between test cases.

Reply:
Ok, I'll add that as part of this PR.

Reply:
Added.